boxtree/fmm.py

```diff
@@ -205,9 +267,46 @@ def drive_fmm(traversal, expansion_wrangler, src_weights, timing_data=None):
 
     # }}}
 
-    result = wrangler.reorder_potentials(potentials)
+    # {{{ Worker ranks send calculated potentials to the root rank
+
+    if distributed:
+        potentials_mpi_type = dtype_to_mpi(potentials.dtype)
+
+        if mpi_rank == 0:
+            potentials_all_ranks = np.empty((mpi_size,), dtype=object)
+            potentials_all_ranks[0] = potentials
 
-    result = wrangler.finalize_potentials(result)
+            for irank in range(1, mpi_size):
+                potentials_all_ranks[irank] = np.empty(
+                    tgt_idx_all_ranks[irank].shape, dtype=potentials.dtype
+                )
+
+                comm.Recv([potentials_all_ranks[irank], potentials_mpi_type],
+                        source=irank, tag=MPITags["GATHER_POTENTIALS"])
```

Comment:
Could do `Irecv` here, for a potential (but probably minor) efficiency gain, to allow the potentials to be received in any order.

Reply:
Implemented [here](https://github.com/inducer/boxtree/commit/500a89b73699778316eecb0c45eb7a95f3c0d2bd#diff-275cfbec0c6196ec964872604af2f4a9db3ca2877591b78750d2e9142da1c805R278-R282).